hooks/pre_gen_project.py

```diff
@@ -8,9 +8,12 @@
 logging.basicConfig(level=logging.DEBUG)
 logger = logging.getLogger("pre_gen_project")
 
-
 if not re.match(r"^[a-z][_a-z0-9]+$", "{{cookiecutter.module_name}}"):
     link = "https://www.python.org/dev/peps/pep-0008/#package-and-module-names"
     logger.error("Module name should be pep-8 compliant.")
     logger.error("  More info: {}".format(link))
     sys.exit(1)
+    
+if re.match(r"_", "{{cookiecutter.module_name}}"):
+    logger.error("PyPI.org and pip do not support package names with underscores.")
```

Comment:
"discourage"? instead of do not support?  there are certainly packages with underscores ([example](https://pypi.org/project/django_testdriver/))... so I wonder if this might mislead someone ever so slightly?

Reply:
interesting. yeah, "discourage" might be better.

I'm struggling to find the details, but it appears that since ~2015, PyPI and setuptools have normalized package names to hyphens: https://www.python.org/dev/peps/pep-0503/#normalized-names

so [pip internally normalizes to dashes](https://pip.pypa.io/en/stable/news/#id44) and PyPI normalizes to dashes (see https://pypi.org/project/napari-time-series-plotter/ and the difference between the PyPI metadata and what the dev wrote in the description)

I wonder if that example has underscores simply because it's pre-2015?

So yeah, "discourage" makes sense.

Reply:
yeah, I also struggled to find what felt like an authoritative explanation of that.  so, 🤷 ?  anyway, we're blocking it anyway, which is the important part here :)